include/swift/AST/LocalizationFormat.h

```diff
@@ -28,6 +31,85 @@ namespace swift {
 enum class DiagID : uint32_t;
 
 namespace diag {
+using namespace llvm::support;
+
+class LocalizationWriterInfo {
+public:
+  using key_type = llvm::StringRef;
+  using key_type_ref = key_type;
+  using data_type = llvm::StringRef;
+  using data_type_ref = data_type;
+  using hash_value_type = uint32_t;
+  using offset_type = uint32_t;
+
+  hash_value_type ComputeHash(key_type_ref key) { return llvm::djbHash(key); }
+
+  std::pair<unsigned, unsigned> EmitKeyDataLength(llvm::raw_ostream &out,
+                                                  key_type_ref key,
+                                                  data_type_ref data) {
+    offset_type keyLength = static_cast<uint32_t>(key.size());
+    offset_type dataLength = static_cast<uint32_t>(data.size());
+    endian::write<offset_type>(out, keyLength, little);
+    endian::write<offset_type>(out, dataLength, little);
+    return {keyLength, dataLength};
+  }
+
+  void EmitKey(llvm::raw_ostream &out, key_type_ref key, unsigned len) {
+    out << key;
+  }
+
+  void EmitData(llvm::raw_ostream &out, key_type_ref key, data_type_ref data,
+                unsigned len) {
+    out << data;
+  }
+};
+
+class LocalizationReaderInfo {
+public:
+  using internal_key_type = llvm::StringRef;
+  using external_key_type = internal_key_type;
+  using data_type = llvm::StringRef;
+  using hash_value_type = uint32_t;
+  using offset_type = uint32_t;
+
+  internal_key_type GetInternalKey(external_key_type key) { return key; }
+
+  external_key_type GetExternalKey(internal_key_type key) { return key; }
+
+  static bool EqualKey(internal_key_type lhs, internal_key_type rhs) {
+    return lhs == rhs;
+  }
+
+  hash_value_type ComputeHash(internal_key_type key) {
+    return llvm::djbHash(key);
+  }
+
+  static std::pair<unsigned, unsigned> ReadKeyDataLength(const uint8_t *&data) {
+    offset_type keyLength =
+        endian::readNext<offset_type, little, unaligned>(data);
+    offset_type dataLength =
+        endian::readNext<offset_type, little, unaligned>(data);
+    return {keyLength, dataLength};
+  }
+
+  internal_key_type ReadKey(const uint8_t *data, offset_type length) {
+    return internal_key_type((const char *)data, length);
+  }
+
+  data_type ReadData(llvm::StringRef Key, const uint8_t *data,
+                     offset_type length) {
+    return data_type((const char *)data, length);
+  }
+};
+
+class SerializedLocalizationWriter {
+  llvm::OnDiskChainedHashTableGenerator<LocalizationWriterInfo> generator;
+
+public:
+  void insert(llvm::StringRef id, llvm::StringRef translation);
+  void emit(llvm::StringRef filePath);
```

Comment:
Let's make `emit` return `bool` to indicate whether translations have been successfully emitted into the file.